src/lstm_layer.cpp

```diff
@@ -1515,3 +1515,505 @@ void LSTM::preinit_layer() {
+
+        lstm_hidden_state_mean_var_mp(
+            lstm_states.mu_o_ga, lstm_states.var_o_ga, lstm_states.mu_ca,
+            lstm_states.var_ca, lstm_states.cov_o_tanh_c, this->output_size,
+            this->seq_len, batch_size, this->num_threads, output_states.mu_a,
+            output_states.var_a);
+
+    } else {
+        lstm_cov_input_cell_states(
+            lstm_states.var_ha, this->mu_w, lstm_states.jcb_i_ga,
+            lstm_states.jcb_c_ga, this->w_pos_i, this->w_pos_c,
+            this->input_size, this->output_size, this->seq_len, batch_size,
+            lstm_states.cov_i_c);
+
+        lstm_cell_state_mean_var(
+            lstm_states.mu_f_ga, lstm_states.var_f_ga, lstm_states.mu_i_ga,
+            lstm_states.var_i_ga, lstm_states.mu_c_ga, lstm_states.var_c_ga,
+            lstm_states.mu_c_prev, lstm_states.var_c_prev, lstm_states.cov_i_c,
+            this->output_size, this->seq_len, batch_size, lstm_states.mu_c,
+            lstm_states.var_c);
+
+        tanh_mean_var(lstm_states.mu_c, lstm_states.var_c, 0, end_chunk,
+                      lstm_states.mu_ca, lstm_states.jcb_ca,
+                      lstm_states.var_ca);
+
+        lstm_cov_output_tanh_cell_states(
+            this->mu_w, lstm_states.var_ha, lstm_states.mu_c_prev,
+            lstm_states.jcb_ca, lstm_states.jcb_f_ga, lstm_states.mu_i_ga,
+            lstm_states.jcb_i_ga, lstm_states.mu_c_ga, lstm_states.jcb_c_ga,
+            lstm_states.jcb_o_ga, this->w_pos_f, this->w_pos_i, this->w_pos_c,
+            this->w_pos_o, this->input_size, this->output_size, this->seq_len,
+            batch_size, lstm_states.cov_o_tanh_c);
+
+        lstm_hidden_state_mean_var(
+            lstm_states.mu_o_ga, lstm_states.var_o_ga, lstm_states.mu_ca,
+            lstm_states.var_ca, lstm_states.cov_o_tanh_c, this->output_size,
+            this->seq_len, batch_size, output_states.mu_a, output_states.var_a);
+    }
+
+    if (this->training) {
+        this->storing_states_for_training(input_states, output_states);
+    }
+    // Save the previous states
+    if (this->seq_len == 1 && batch_size == 1) {
+        lstm_to_prev_states(output_states.mu_a,
+                            this->lstm_states.mu_h_prior.size(),
+                            this->lstm_states.mu_h_prior);
+        lstm_to_prev_states(output_states.var_a,
+                            this->lstm_states.var_h_prior.size(),
+                            this->lstm_states.var_h_prior);
+        lstm_to_prev_states(this->lstm_states.mu_c,
+                            this->lstm_states.mu_c_prior.size(),
+                            this->lstm_states.mu_c_prior);
+        lstm_to_prev_states(this->lstm_states.var_c,
+                            this->lstm_states.var_c_prior.size(),
+                            this->lstm_states.var_c_prior);
+
+        // Save priors for smoothing
+        this->lstm_states.mu_h_priors.push_back(this->lstm_states.mu_h_prior);
```

Comment:
Avoid 2d vector and push_back. `push_back` will cost the performance.

Reply:
In order to allocate a vector, you will need to know the exact size of that vector. Otherwise, it is gonna be slow for cuda version where we need to reallocate the memory every time

Reply:
I added num_sample = number_timestep to Sequential class to have enough information for initializing vectors.